File: backend/src/services/aiCampaignTools.ts
import {
  resolveLocalizedText,
  validateFormDefinition,
  type AICampaignReference,
  type QuestionDefinition,
  type ValidationResult,
} from "@formbuilder/shared";
import type {
  SearchCampaignsArgs,
  GetCampaignArgs,
  GetCampaignQuestionsArgs,
  SearchQuestionsArgs,
  FindSimilarCampaignsArgs,
  FindSimilarQuestionsArgs,
  ValidateFormArgs,
} from "@formbuilder/shared";
import type { UserRole } from "../entities/User";
import { isAdminRole } from "../entities/User";
import { getFormDetail, listForms, type FormDetail, type FormListItem } from "./formBuilderService";
import { getAccessibleFormDetail, listAccessibleForms } from "./formAccessService";

/**
 * Read-only tool implementations backing the AI assistant's `READ_ONLY_AI_TOOLS`
 * (see @formbuilder/shared's aiTypes.ts). Every function here takes the
 * calling user's own auth context and enforces the *exact* same
 * subsidiary-scoping subsidiaryForms.router.ts already applies for a human
 * hitting these same underlying services directly — a non-admin's
 * `subsidiaryId` always comes from this context, never from a model-supplied
 * tool argument (there is no subsidiaryId argument on any of these tool
 * shapes in the first place, precisely so nothing can be trusted from the
 * model here).
 *
 * `GET_CAMPAIGN`/`GET_CAMPAIGN_QUESTIONS` intentionally return a *compact*
 * projection (name/status/locales/questions reduced to id+heading+type+
 * required) rather than the full FormDefinition JSON — keeps token cost down
 * and avoids ever handing the model profile-field/validation-message/consent
 * content it has no reason to see.
 *
 * `SEARCH_QUESTIONS`/`FIND_SIMILAR_CAMPAIGNS`/`FIND_SIMILAR_QUESTIONS` are
 * deliberately simple v1 in-memory keyword/token-overlap scans over a capped
 * number of accessible forms — not a real search index or embeddings. This
 * is a conscious v1 choice (per the plan) with a stable calling contract, so
 * it can be swapped for vector search later without any caller needing to
 * change.
 */

export interface AiToolCallerContext {
  userId: string;
  role: UserRole;
  subsidiaryId: string | null;
}

/** Cap on how many forms a keyword scan reads full content for — this is an
 * in-memory O(N) scan, not an index, so N must stay small enough to run
 * inline within a chat request. */
const SCAN_FORM_LIMIT = 100;
/** Cap on how many results any of the "search"/"similar" tools returns to the
 * model — keeps the tool-result turn's token cost bounded regardless of how
 * many forms exist. */
const RESULT_LIMIT = 10;

export interface CompactQuestion {
  id: string;
  heading: string;
  controlType: string;
  required: boolean;
}

export interface CompactCampaign {
  formId: string;
  name: string;
  status: string;
  locales: string[];
  defaultLocale: string;
  questions: CompactQuestion[];
}

export interface CampaignSearchResult {
  formId: string;
  name: string;
  status: string;
  projectCode: string | null;
  questionCount: number;
  updatedAt: string;
}

export interface QuestionSearchResult {
  formId: string;
  formName: string;
  questionId: string;
  heading: string;
  controlType: string;
}

/** Every form the caller may read, scoped exactly like a human hitting
 * GET /api/v1/admin/forms (admin) or GET /api/v1/forms (subsidiary user)
 * would see — admins see every non-deleted form regardless of subsidiary,
 * a subsidiary-scoped standard user sees only their own subsidiary's
 * *published* forms (formAccessService's own restriction, not loosened
 * here), and a standard user with no subsidiary assigned sees nothing. */
async function listCallerForms(
  ctx: AiToolCallerContext,
  filters: { search?: string; projectCode?: string; status?: SearchCampaignsArgs["status"] } = {},
): Promise<FormListItem[]> {
  if (isAdminRole(ctx.role)) {
    const page = await listForms({
      search: filters.search,
      projectCode: filters.projectCode,
      status: filters.status,
      pageSize: SCAN_FORM_LIMIT,
    });
    return page.items;
  }

  if (!ctx.subsidiaryId) return [];
  let forms = await listAccessibleForms(ctx.subsidiaryId, ctx.userId);
  if (filters.status) forms = forms.filter((f) => f.status === filters.status);
  if (filters.projectCode) forms = forms.filter((f) => f.projectCode === filters.projectCode);
  if (filters.search) {
    const needle = filters.search.toLowerCase();
    forms = forms.filter((f) => f.name.toLowerCase().includes(needle));
  }
  return forms.slice(0, SCAN_FORM_LIMIT);
}

/** Same access rule as `listCallerForms`, for a single form by id — returns
 * null identically whether the form doesn't exist or the caller can't see
 * it, matching this codebase's "never leak which case occurred" ownership
 * convention (see uploadService.findOwnedUpload's own doc comment). */
export async function getCallerFormDetail(ctx: AiToolCallerContext, formId: string): Promise<FormDetail | null> {
  if (isAdminRole(ctx.role)) return getFormDetail(formId);
  if (!ctx.subsidiaryId) return null;
  return getAccessibleFormDetail(formId, ctx.subsidiaryId);
}

function toCompactCampaign(formId: string, name: string, status: string, detail: FormDetail): CompactCampaign {
  const content = detail.draft ?? detail.published;
  if (!content) {
    return { formId, name, status, locales: [], defaultLocale: "", questions: [] };
  }
  const { definition } = content;
  return {
    formId,
    name,
    status,
    locales: definition.locales.map((l) => l.code),
    defaultLocale: definition.meta.defaultLocale,
    questions: definition.questions
      .slice()
      .sort((a, b) => a.order - b.order)
      .map((q) => ({
        id: q.id,
        heading: resolveLocalizedText(q.headingByLocale, definition.meta.defaultLocale, definition.meta.defaultLocale),
        controlType: q.controlType,
        required: q.required,
      })),
  };
}

export async function searchCampaigns(ctx: AiToolCallerContext, args: SearchCampaignsArgs): Promise<CampaignSearchResult[]> {
  const forms = await listCallerForms(ctx, { search: args.searchText, projectCode: args.projectCode, status: args.status });
  return forms.slice(0, RESULT_LIMIT).map((f) => ({
    formId: f.id,
    name: f.name,
    status: f.status,
    projectCode: f.projectCode,
    questionCount: 0, // list items don't carry parsed content; see getCampaign for a full question list
    updatedAt: f.updatedAt.toISOString(),
  }));
}

export async function getCampaign(ctx: AiToolCallerContext, args: GetCampaignArgs): Promise<CompactCampaign | null> {
  const detail = await getCallerFormDetail(ctx, args.formId);
  if (!detail) return null;
  return toCompactCampaign(detail.id, detail.name, detail.status, detail);
}

export async function getCampaignQuestions(ctx: AiToolCallerContext, args: GetCampaignQuestionsArgs): Promise<CompactQuestion[] | null> {
  const campaign = await getCampaign(ctx, { formId: args.formId });
  return campaign ? campaign.questions : null;
}

/** Lowercased, punctuation-stripped word set — the unit both the "search" and
 * "similar" keyword scans compare on. Deliberately naive (no stemming/
 * stopwords) — see this file's own doc comment on why that's fine for v1. */
function tokenize(text: string): Set<string> {
  return new Set(
    text
      .toLowerCase()
      .split(/[^a-z0-9]+/i)
      .filter((t) => t.length > 1),
  );
}

function overlapScore(a: Set<string>, b: Set<string>): number {
  let score = 0;
  for (const token of a) if (b.has(token)) score++;
  return score;
}

export async function searchQuestions(ctx: AiToolCallerContext, args: SearchQuestionsArgs): Promise<QuestionSearchResult[]> {
  const forms = args.formId
    ? [await getCallerFormDetail(ctx, args.formId)].filter((f): f is FormDetail => f !== null)
    : await Promise.all((await listCallerForms(ctx)).map((f) => getCallerFormDetail(ctx, f.id))).then((rows) =>
        rows.filter((f): f is FormDetail => f !== null),
      );

  const needle = args.searchText.toLowerCase();
  const results: QuestionSearchResult[] = [];
  for (const form of forms) {
    const content = form.draft ?? form.published;
    if (!content) continue;
    for (const q of content.definition.questions) {
      const heading = resolveLocalizedText(q.headingByLocale, content.definition.meta.defaultLocale, content.definition.meta.defaultLocale);
      if (heading.toLowerCase().includes(needle)) {
        results.push({ formId: form.id, formName: form.name, questionId: q.id, heading, controlType: q.controlType });
      }
    }
    if (results.length >= RESULT_LIMIT) break;
  }
  return results.slice(0, RESULT_LIMIT);
}

export async function findSimilarCampaigns(ctx: AiToolCallerContext, args: FindSimilarCampaignsArgs): Promise<CampaignSearchResult[]> {
  const target = await getCallerFormDetail(ctx, args.formId);
  if (!target) return [];
  const targetTokens = tokenize(target.name);

  const candidates = await listCallerForms(ctx);
  const scored = candidates
    .filter((f) => f.id !== args.formId)
    .map((f) => ({ form: f, score: overlapScore(targetTokens, tokenize(f.name)) }))
    .filter((s) => s.score > 0)
    .sort((a, b) => b.score - a.score)
    .slice(0, RESULT_LIMIT);

  return scored.map(({ form }) => ({
    formId: form.id,
    name: form.name,
    status: form.status,
    projectCode: form.projectCode,
    questionCount: 0,
    updatedAt: form.updatedAt.toISOString(),
  }));
}

export async function findSimilarQuestions(ctx: AiToolCallerContext, args: FindSimilarQuestionsArgs): Promise<QuestionSearchResult[]> {
  let targetText = args.text ?? "";
  if (!targetText && args.formId && args.questionId) {
    const campaign = await getCampaign(ctx, { formId: args.formId });
    targetText = campaign?.questions.find((q) => q.id === args.questionId)?.heading ?? "";
  }
  if (!targetText) return [];
  const targetTokens = tokenize(targetText);

  const forms = await Promise.all((await listCallerForms(ctx)).map((f) => getCallerFormDetail(ctx, f.id))).then((rows) =>
    rows.filter((f): f is FormDetail => f !== null),
  );

  const results: { result: QuestionSearchResult; score: number }[] = [];
  for (const form of forms) {
    const content = form.draft ?? form.published;
    if (!content) continue;
    for (const q of content.definition.questions) {
      if (args.formId && args.questionId && form.id === args.formId && q.id === args.questionId) continue;
      const heading = resolveLocalizedText(q.headingByLocale, content.definition.meta.defaultLocale, content.definition.meta.defaultLocale);
      const score = overlapScore(targetTokens, tokenize(heading));
      if (score > 0) {
        results.push({ result: { formId: form.id, formName: form.name, questionId: q.id, heading, controlType: q.controlType }, score });
      }
    }
  }
  return results
    .sort((a, b) => b.score - a.score)
    .slice(0, RESULT_LIMIT)
    .map((r) => r.result);
}

export async function validateForm(ctx: AiToolCallerContext, args: ValidateFormArgs): Promise<ValidationResult | null> {
  const detail = await getCallerFormDetail(ctx, args.formId);
  const content = detail?.draft ?? detail?.published;
  if (!content) return null;
  return validateFormDefinition(content.definition);
}

/** Builds `AIChatResponse.references` for a SEARCH_CAMPAIGNS/FIND_SIMILAR_CAMPAIGNS
 * tool result — fetches each candidate's full detail (bounded by the RESULT_LIMIT-sized
 * input list, so at most a handful of extra reads) to fill in `locales`/`questionCount`,
 * which the plain search-result rows above don't carry. Silently drops a formId that's
 * no longer accessible (e.g. unpublished between the search and this call) rather than
 * failing the whole response. */
export async function buildCampaignReferences(ctx: AiToolCallerContext, results: CampaignSearchResult[]): Promise<AICampaignReference[]> {
  const details = await Promise.all(results.map((r) => getCallerFormDetail(ctx, r.formId)));
  const references: AICampaignReference[] = [];
  for (let i = 0; i < results.length; i++) {
    const detail = details[i];
    if (!detail) continue;
    const content = detail.draft ?? detail.published;
    references.push({
      formId: results[i].formId,
      name: results[i].name,
      status: results[i].status,
      questionCount: content?.definition.questions.length ?? 0,
      locales: content?.definition.locales.map((l) => l.code) ?? [],
      updatedAt: results[i].updatedAt,
    });
  }
  return references;
}

/** Referenced by aiAssistantService.ts when it needs a QuestionDefinition's
 * heading purely for logging/echoing purposes — kept here alongside the
 * other question-shaped helpers rather than duplicated. */
export function questionHeading(question: QuestionDefinition, defaultLocale: string): string {
  return resolveLocalizedText(question.headingByLocale, defaultLocale, defaultLocale);
}

File: backend-py/app/services/aiCampaignTools.py
"""Port of `backend/src/services/aiCampaignTools.ts`.

Read-only tool implementations backing the AI assistant. Every function enforces
the same subsidiary-scoping as subsidiaryForms.router.ts.
"""

from __future__ import annotations

import re
from typing import Any, Optional, TypedDict

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.form_pipeline import resolve_localized_text, validate_form_definition
from app.models.form import Form
from app.models.user import is_admin_role
from app.services import form_builder_service
from app.services.form_access_service import get_accessible_form_detail, list_accessible_forms

SCAN_FORM_LIMIT = 100
RESULT_LIMIT = 10


class AiToolCallerContext(TypedDict):
    userId: str
    role: str
    subsidiaryId: Optional[str]


class CompactQuestion(TypedDict):
    id: str
    heading: str
    controlType: str
    required: bool


class CompactCampaign(TypedDict):
    formId: str
    name: str
    status: str
    locales: list[str]
    defaultLocale: str
    questions: list[CompactQuestion]


class CampaignSearchResult(TypedDict):
    formId: str
    name: str
    status: str
    projectCode: Optional[str]
    questionCount: int
    updatedAt: str


class QuestionSearchResult(TypedDict):
    formId: str
    formName: str
    questionId: str
    heading: str
    controlType: str


def _list_caller_forms(
    db: Session, ctx: AiToolCallerContext,
    *, search: Optional[str] = None, project_code: Optional[str] = None,
    status: Optional[str] = None,
) -> list[dict[str, Any]]:
    """Every form the caller may read, scoped by role."""
    if is_admin_role(ctx["role"]):
        page = form_builder_service.list_forms(
            db, page_size=SCAN_FORM_LIMIT, search=search,
            project_code=project_code, status=status,
        )
        return page["items"]

    if not ctx["subsidiaryId"]:
        return []

    accessible = list_accessible_forms(db, ctx["subsidiaryId"], ctx["userId"])
    my_adhoc = form_builder_service.list_my_adhoc_forms(db, ctx["subsidiaryId"])
    by_id = {f["id"]: f for f in accessible}
    for f in my_adhoc:
        if f["id"] not in by_id:
            by_id[f["id"]] = f

    # Include drafts
    draft_rows = list(
        db.execute(
            select(Form).where(
                Form.subsidiaryId == ctx["subsidiaryId"],
                Form.status == "draft",
                Form.isDeleted == False,  # noqa: E712
            ).order_by(Form.updatedAt.desc())
        ).scalars()
    )
    for f in draft_rows:
        fid = f.id
        if fid not in by_id:
            by_id[fid] = {
                "id": f.id, "name": f.name, "subsidiaryId": f.subsidiaryId,
                "projectCode": f.projectCode, "status": f.status,
                "createdByUserId": f.createdByUserId, "createdAt": f.createdAt,
                "updatedAt": f.updatedAt, "publishedVersionNumber": None,
                "origin": f.origin, "pendingReview": f.pendingReview,
                "submittedForReviewAt": f.submittedForReviewAt,
                "reviewedAt": f.reviewedAt, "reviewNote": f.reviewNote,
            }

    forms = list(by_id.values())
    if status:
        forms = [f for f in forms if f["status"] == status]
    if project_code:
        forms = [f for f in forms if f.get("projectCode") == project_code]
    if search:
        needle = search.lower()
        forms = [f for f in forms if needle in f["name"].lower()]
    return forms[:SCAN_FORM_LIMIT]


def get_caller_form_detail(
    db: Session, ctx: AiToolCallerContext, form_id: str
) -> Optional[dict[str, Any]]:
    """Same access rule as _list_caller_forms, for a single form by id."""
    if is_admin_role(ctx["role"]):
        return form_builder_service.get_form_detail(db, form_id)
    if not ctx["subsidiaryId"]:
        return None

    accessible = get_accessible_form_detail(db, form_id, ctx["subsidiaryId"])
    if accessible:
        return accessible

    owned = form_builder_service.find_owned_adhoc_form(db, form_id, ctx["subsidiaryId"])
    if owned:
        return form_builder_service.get_form_detail(db, form_id)
    return None


def _to_compact_campaign(
    form_id: str, name: str, status: str, detail: dict[str, Any]
) -> CompactCampaign:
    content = detail.get("draft") or detail.get("published")
    if not content:
        return {"formId": form_id, "name": name, "status": status, "locales": [], "defaultLocale": "", "questions": []}
    definition = content["definition"]
    default_locale = definition.meta.defaultLocale
    questions = sorted(definition.questions, key=lambda q: q.order)
    return {
        "formId": form_id,
        "name": name,
        "status": status,
        "locales": [l.code for l in definition.locales],
        "defaultLocale": default_locale,
        "questions": [
            {
                "id": q.id,
                "heading": resolve_localized_text(q.headingByLocale, default_locale, default_locale),
                "controlType": q.controlType,
                "required": q.required,
            }
            for q in questions
        ],
    }


def _normalize(text: str) -> str:
    return re.sub(r"[^a-z0-9]+", "", text.lower())


def _tokenize(text: str) -> set[str]:
    return {t for t in re.split(r"[^a-z0-9]+", text.lower()) if len(t) > 1}


def _overlap_score(a: set[str], b: set[str]) -> int:
    return sum(1 for t in a if t in b)


def search_campaigns(
    db: Session, ctx: AiToolCallerContext, args: dict[str, Any]
) -> list[CampaignSearchResult]:
    search_text = args.get("searchText")
    project_code = args.get("projectCode")
    status = args.get("status")

    forms = _list_caller_forms(
        db, ctx, search=search_text, project_code=project_code, status=status
    )

    # Fallback: normalize and tokenize if no results
    if not forms and search_text:
        candidates = _list_caller_forms(db, ctx, project_code=project_code, status=status)
        needle = _normalize(search_text)
        forms = [f for f in candidates if needle in _normalize(f["name"])]

        if not forms:
            needle_tokens = _tokenize(search_text)
            scored = [
                (f, _overlap_score(needle_tokens, _tokenize(f["name"])))
                for f in candidates
            ]
            scored = [(f, s) for f, s in scored if s > 0]
            scored.sort(key=lambda x: x[1], reverse=True)
            forms = [f for f, _ in scored]

    results: list[CampaignSearchResult] = []
    for f in forms[:RESULT_LIMIT]:
        results.append({
            "formId": f["id"],
            "name": f["name"],
            "status": f["status"],
            "projectCode": f.get("projectCode"),
            "questionCount": 0,
            "updatedAt": f["updatedAt"].isoformat() if hasattr(f["updatedAt"], "isoformat") else str(f["updatedAt"]),
        })
    return results


def get_campaign(
    db: Session, ctx: AiToolCallerContext, args: dict[str, Any]
) -> Optional[CompactCampaign]:
    detail = get_caller_form_detail(db, ctx, args["formId"])
    if not detail:
        return None
    return _to_compact_campaign(detail["id"], detail["name"], detail["status"], detail)


def get_campaign_questions(
    db: Session, ctx: AiToolCallerContext, args: dict[str, Any]
) -> Optional[list[CompactQuestion]]:
    campaign = get_campaign(db, ctx, args)
    return campaign["questions"] if campaign else None


def search_questions(
    db: Session, ctx: AiToolCallerContext, args: dict[str, Any]
) -> list[QuestionSearchResult]:
    form_id = args.get("formId")
    search_text = args["searchText"]
    needle = search_text.lower()

    if form_id:
        detail = get_caller_form_detail(db, ctx, form_id)
        forms = [detail] if detail else []
    else:
        form_list = _list_caller_forms(db, ctx)
        forms = []
        for f in form_list:
            detail = get_caller_form_detail(db, ctx, f["id"])
            if detail:
                forms.append(detail)

    results: list[QuestionSearchResult] = []
    for form in forms:
        content = form.get("draft") or form.get("published")
        if not content:
            continue
        definition = content["definition"]
        default_locale = definition.meta.defaultLocale
        for q in definition.questions:
            heading = resolve_localized_text(q.headingByLocale, default_locale, default_locale)
            if needle in heading.lower():
                results.append({
                    "formId": form["id"],
                    "formName": form["name"],
                    "questionId": q.id,
                    "heading": heading,
                    "controlType": q.controlType,
                })
            if len(results) >= RESULT_LIMIT:
                break
        if len(results) >= RESULT_LIMIT:
            break
    return results[:RESULT_LIMIT]


def find_similar_campaigns(
    db: Session, ctx: AiToolCallerContext, args: dict[str, Any]
) -> list[CampaignSearchResult]:
    target = get_caller_form_detail(db, ctx, args["formId"])
    if not target:
        return []
    target_tokens = _tokenize(target["name"])

    candidates = _list_caller_forms(db, ctx)
    scored = [
        (f, _overlap_score(target_tokens, _tokenize(f["name"])))
        for f in candidates if f["id"] != args["formId"]
    ]
    scored = [(f, s) for f, s in scored if s > 0]
    scored.sort(key=lambda x: x[1], reverse=True)

    results: list[CampaignSearchResult] = []
    for f, _ in scored[:RESULT_LIMIT]:
        results.append({
            "formId": f["id"],
            "name": f["name"],
            "status": f["status"],
            "projectCode": f.get("projectCode"),
            "questionCount": 0,
            "updatedAt": f["updatedAt"].isoformat() if hasattr(f["updatedAt"], "isoformat") else str(f["updatedAt"]),
        })
    return results


def find_similar_questions(
    db: Session, ctx: AiToolCallerContext, args: dict[str, Any]
) -> list[QuestionSearchResult]:
    target_text = args.get("text", "")
    if not target_text and args.get("formId") and args.get("questionId"):
        campaign = get_campaign(db, ctx, {"formId": args["formId"]})
        if campaign:
            q = next((q for q in campaign["questions"] if q["id"] == args["questionId"]), None)
            if q:
                target_text = q["heading"]
    if not target_text:
        return []

    target_tokens = _tokenize(target_text)
    form_list = _list_caller_forms(db, ctx)
    details = []
    for f in form_list:
        d = get_caller_form_detail(db, ctx, f["id"])
        if d:
            details.append(d)

    scored: list[tuple[QuestionSearchResult, int]] = []
    for form in details:
        content = form.get("draft") or form.get("published")
        if not content:
            continue
        definition = content["definition"]
        default_locale = definition.meta.defaultLocale
        for q in definition.questions:
            if args.get("formId") and args.get("questionId"):
                if form["id"] == args["formId"] and q.id == args["questionId"]:
                    continue
            heading = resolve_localized_text(q.headingByLocale, default_locale, default_locale)
            score = _overlap_score(target_tokens, _tokenize(heading))
            if score > 0:
                scored.append((
                    {
                        "formId": form["id"],
                        "formName": form["name"],
                        "questionId": q.id,
                        "heading": heading,
                        "controlType": q.controlType,
                    },
                    score,
                ))

    scored.sort(key=lambda x: x[1], reverse=True)
    return [r for r, _ in scored[:RESULT_LIMIT]]


def validate_form(
    db: Session, ctx: AiToolCallerContext, args: dict[str, Any]
) -> Optional[dict[str, Any]]:
    detail = get_caller_form_detail(db, ctx, args["formId"])
    if not detail:
        return None
    content = detail.get("draft") or detail.get("published")
    if not content:
        return None
    validation = validate_form_definition(content["definition"])
    return {"errors": validation.errors, "warnings": validation.warnings}


async def build_campaign_references(
    db: Session, ctx: AiToolCallerContext, results: list[CampaignSearchResult]
) -> list[dict[str, Any]]:
    """Builds AIChatResponse.references for SEARCH_CAMPAIGNS/FIND_SIMILAR_CAMPAIGNS."""
    references = []
    for r in results:
        detail = get_caller_form_detail(db, ctx, r["formId"])
        if not detail:
            continue
        content = detail.get("draft") or detail.get("published")
        references.append({
            "formId": r["formId"],
            "name": r["name"],
            "status": r["status"],
            "origin": detail.get("origin", "admin"),
            "questionCount": len(content["definition"].questions) if content else 0,
            "locales": [l.code for l in content["definition"].locales] if content else [],
            "updatedAt": r["updatedAt"],
        })
    return references
